docs/src/start/logging-packets.md

```diff
@@ -0,0 +1,311 @@
+
+### Writing Data To The Map
+
+With our helper function in place, we can:
+1. Read the Ethertype field to check if we have an IPv4 packet.
+1. Read the IPv4 Source Address from the IP header
+
+First let's add another dependency on `memoffset = "0.6"` to `myapp-ebpf/Cargo.toml`, and then we'll change our `try_xdp_firewall` function to look like this:
+
+```rust,ignore
+fn try_xdp_firewall(ctx: XdpContext) -> Result<u32, ()> {
+    let h_proto = u16::from_be(unsafe { *ptr_at(&ctx, offset_of!(ethhdr, h_proto))? });
+    if h_proto != ETH_P_IP {
+        return Ok(xdp_action::XDP_PASS)
+    }
+    let source = u32::from_be(unsafe { *ptr_at(&ctx, ETH_HDR_LEN + offset_of!(iphdr, saddr))? });
+
+    let log_entry = PacketLog{
+        ipv4_address: source,
+        action: xdp_action::XDP_PASS,
+    };
+    unsafe { EVENTS.output(&ctx, &log_entry, 0); }
+    Ok(xdp_action::XDP_PASS)
+}
+```
+
+> 💡 **HINT: Reading Fields Using `offset_of!`**
+>
+> As there is limited stack space, it's more memory efficient to use the `offset-of!` macro to read
+> a single field from a struct, rather than reading the whole struct and accessing the field by name.
+
+Once we have our IPv4 source address, we can create a `PacketLog` struct and output this to our PerfEventArray
+
+## Reading Data
+
+### Going Async
+
+In order to read from the `AsyncPerfEventArray`, we have to call `AsyncPerfEventArray::open()` for each online CPU, then we have to poll the file descriptor for events.
+While this is do-able using `PerfEventArray` and `mio` or `epoll`, the code is much less easy to follow. Instead, we'll use `tokio` to make our user-space application async.
+
+
+Let's add some dependencies to `myapp/src/Cargo.toml`:
+
+```toml
+[dependencies]
+aya = { git = "https://github.com/alessandrod/aya", branch="main", features=["async_tokio"] }
+myapp-common = { path = "../myapp-common", features=["userspace"] }
+anyhow = "1.0.42"
+bytes = "1"
+tokio = { version = "1.9.0", features = ["full"] }
+```
+
+
+And adjust our `myapp/src/main.rs` to look like this:
+
+```rust,ignore
+use aya::{
+    maps::perf::AsyncPerfEventArray,
+    programs::{Xdp, XdpFlags},
+    util::online_cpus,
+    Bpf,
+};
+use bytes::BytesMut;
+use std::{
+    convert::{TryFrom, TryInto},
+    env, fs, net,
+};
+use tokio::{signal, task};
+
+use myapp_common::PacketLog;
+
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
+    let path = match env::args().nth(1) {
+        Some(iface) => iface,
+        None => panic!("not path provided"),
+    };
+    let iface = match env::args().nth(2) {
+        Some(iface) => iface,
+        None => "eth0".to_string(),
+    };
+
+    let data = fs::read(path)?;
+    let mut bpf = Bpf::load(&data, None)?;
+
+    let probe: &mut Xdp = bpf.program_mut("xdp")?.try_into()?;
+    probe.load()?;
+    probe.attach(&iface, XdpFlags::default())?;
```

Comment:
Here we should add a comment that says that if you get an error loading the program, you should try to pass `XdpFlags::SKB_MODE` to attach. Some virtio drivers (eg on digital ocean), don't support hw mode.